python/cudf/cudf/core/dataframe.py

```diff
@@ -6380,71 +6380,294 @@ def to_orc(
+
+        Stacking a dataframe with a multi-level column axis:
+
+        >>> df_multi_level_cols1
+             weight
+                 kg    pounds
+        cat       1        2
+        dog       2        4
+        >>> df_multi_level_cols1.stack()
+                    weight
+        cat kg           1
+            pounds       2
+        dog kg           2
+            pounds       4
+
+        **Missing values**
+
+        >>> multicol2 = pd.MultiIndex.from_tuples([('weight', 'kg'),
+        ...                                        ('height', 'm')])
+        >>> df_multi_level_cols2 = cudf.DataFrame([[1.0, 2.0], [3.0, 4.0]],
+        ...                                     index=['cat', 'dog'],
+        ...                                     columns=multicol2)
+
+        It is common to have missing values when stacking a dataframe
+        with multi-level columns, as the stacked dataframe typically
+        has more values than the original dataframe. Missing values
+        are filled with NULLs:
+
+        >>> df_multi_level_cols2
+            weight height
+                kg      m
+        cat    1.0    2.0
+        dog    3.0    4.0
+        >>> df_multi_level_cols2.stack()
+            height weight
+        cat kg   <NA>    1.0
+            m     2.0   <NA>
+        dog kg   <NA>    3.0
+            m     4.0   <NA>
+
+        **Prescribing the level(s) to be stacked**
+
+        The first parameter controls which level or levels are stacked:
+
+        >>> df_multi_level_cols2.stack(0)
+                    kg     m
+        cat height  <NA>   2.0
+            weight   1.0  <NA>
+        dog height  <NA>   4.0
+            weight   3.0  <NA>
+
+        >>> df_multi_level_cols2.stack([0, 1])
+        cat  height  m     2.0
+             weight  kg    1.0
+        dog  height  m     4.0
+             weight  kg    3.0
+        dtype: float64
         """
-        assert level in (None, -1)
-        repeated_index = self.index.repeat(self.shape[1])
-        name_index = libcudf.reshape.tile(
-            [as_column(self._column_names)], self.shape[0]
+
+        level = [level] if not isinstance(level, list) else level
+
+        if len(level) > 1 and not dropna:
+            raise NotImplementedError(
+                "When stacking multiple levels, setting `dropna` to False "
+                "will generate new column combination that does not exist "
+                "in original dataframe. This behavior is unsupported in "
+                "cuDF. See pandas deprecation note: "
+                "https://github.com/pandas-dev/pandas/issues/53515"
+            )
+
+        # Compute the columns to stack based on specified levels
+        if any(not isinstance(lv, int) for lv in level):
+            # Convert level names to indices, assume a list of labels
+            names = pd.Index(self._data.level_names)
+            level = [names.get_loc(lv) for lv in level]
```

Comment:
You are right here. I was under the shadow of the broad range of label types that pandas allow in the column. If we limit the column names to (int, str) as above, we can avoid calling the pandas helper.

Reply:
There's also the slight concern to that `.index` is O(N), but considering we are promised to work with a short list here the performance should still outperform dispatching to pandas.

Reply:
> This doesn't work if the level list is a mix of int positions and str positions.

Pandas doesn't allow `level` to be specified with mixed integer position and str names as well:
```python

In [24]: midx = pd.MultiIndex.from_arrays([[1, 2, 3], [10, 11, 12]], names=[128, "a"])

In [25]: df = pd.DataFrame(columns=midx)

In [28]: df.stack(level=[0, "a"])

ValueError: level should contain all level names or all level numbers, not a mixture of the two.
```

However the following is allowed in pandas (which doesn't in this PR):

```python
In [17]: midx = pd.MultiIndex.from_arrays([[1, 2, 3], [10, 11, 12]], names=[128, 129])

In [18]: df = pd.DataFrame(columns=midx)

In [20]: df.stack(level=[128])
Out[20]: 
Empty DataFrame
Columns: [10, 11, 12]
Index: []
```

Reply:
>  Since integers can be both a name or a position, pandas always preemptively check if the integer is a column name first, if it doesn't then assumes the integer is a position.

Ah yes, I hate this behaviour...